core/src/main/java/io/parsingdata/metal/expression/value/Scope.java

```diff
@@ -0,0 +1,105 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.parsingdata.metal.expression.value;
+
+import static java.math.BigInteger.ZERO;
+
+import static io.parsingdata.metal.Trampoline.complete;
+import static io.parsingdata.metal.Trampoline.intermediate;
+import static io.parsingdata.metal.Util.checkNotNull;
+import static io.parsingdata.metal.expression.value.NotAValue.NOT_A_VALUE;
+
+import java.util.Objects;
+
+import io.parsingdata.metal.Trampoline;
+import io.parsingdata.metal.Util;
+import io.parsingdata.metal.data.ImmutableList;
+import io.parsingdata.metal.data.ParseGraph;
+import io.parsingdata.metal.data.ParseState;
+import io.parsingdata.metal.encoding.Encoding;
+import io.parsingdata.metal.token.Token;
+
+/**
+ * A {@link ValueExpression} that prunes the {@link ParseState#order} field to
+ * let a nested {@link ValueExpression} have only a scoped view.
+ * <p>
+ * A Scope expression has two operands: <code>scopedValueExpression</code> (a
+ * {@link ValueExpression} and <code>scopeSize</code> (a
+ * {@link SingleValueExpression}). The <code>scopeSize</code> operand is first
+ * evaluated, if it evaluates to <code>NOT_A_VALUE</code> or a negative value
+ * an {@link IllegalArgumentException} is thrown. Otherwise the
+ * <code>order</code> field of the <code>parseState</code> argument is pruned.
+ * This is done by removing all enclosing scopes above the amount specified by
+ * <code>scopeSize</code>. An enclosing scope is defined as an unclosed branch
+ * in the {@link ParseGraph} that has a scope delimiting {@link Token} as its
+ * <code>definition</code>. Next, the <code>scopedValueExpression</code> is
+ * evaluated with this pruned {@link ParseGraph} in the {@link ParseState}.
+ */
+public class Scope implements ValueExpression {
+
+    public final ValueExpression scopedValueExpression;
+    public final SingleValueExpression scopeSize;
+
+    public Scope(final ValueExpression scopedValueExpression, final SingleValueExpression scopeSize) {
+        this.scopedValueExpression = checkNotNull(scopedValueExpression, "scopedValueExpression");
+        this.scopeSize = checkNotNull(scopeSize, "scopeSize");
+    }
+
+    @Override
+    public ImmutableList<Value> eval(final ParseState parseState, final Encoding encoding) {
+        return scopeSize.evalSingle(parseState, encoding)
+            .filter(sizeValue -> !sizeValue.equals(NOT_A_VALUE) && sizeValue.asNumeric().compareTo(ZERO) >= 0)
+            .map(sizeValue -> scopedValueExpression.eval(parseState.withOrder(calculateScope(parseState.order, sizeValue.asNumeric().intValueExact())), encoding))
+            .orElseThrow(() -> new IllegalArgumentException("Argument scopeSize must evaluate to a positive, non-empty countable value."));
+
+    }
+
+    private ParseGraph calculateScope(final ParseGraph order, final int size) {
```

Comment:
Is `calculate` the right verb here? Maybe use `determine` or `select`?

Reply:
I was thinking about `createScope`, but we already have `createScopeList` and `findScope` further down. I agree _calculate_ is a bit dodgy, but I tried to capture the intuition of having a graph, picking out parts and then determine which one is needed. I think _select_ matches that the best but it's a little ambiguous with _find_. _Calculate_ at least communicates that it's the thing that comes up with the final answer.